README.md
# django-rest-framework-httpsignature

TODO:
    * Travis tag

## Overview

Provides [HTTP Signature](http://tools.ietf.org/html/draft-cavage-http-signatures-01) support for [Django REST framework](http://django-rest-framework.org/). The HTTP Signature provides a way to achieve origin authentication and message integrity to HTTP messages.

TODO:
    * Mention Amazon [HTTP Signature scheme](http://docs.aws.amazon.com/general/latest/gr/signature-version-4.html), and mention main differences.

## Installation

```
pip install git+https://github.com/etoccalino/django-rest-framework-httpsignature
```

## Usage

TODO:
    * Extend and complete SingatureAuthentication.
    * Configure REST_FRAMEWORK in settings.py

## Example usage & session w/cURL

TODO:
    * extending SignatureAuthentication for in-memory api key data store.
    * edit setting.py to require authentication.
    * using cURL to hit a URL.


CHANGES.txt
v1.0.0, 20150501 -- Relaxed dependencies

        * Let Django version float all the way up to 1.8.x and above
          Thanks Markus Holtermann
        * Let DRF version float all the way up to 3.x and above
          Thanks Sofia Margariti

v0.2.1, 20140818 -- Fix tests

        * Update to latest version of httpsig
        * Brought 0.1.x fixes in to 0.2.x branch

v0.2.0, 20140803 -- Move to use httpsig package

        * Update documentation
        * Adapt to httpsig (thanks Konstantinos Koukopoulos)
        * Using draft 03 lastest version

v0.1.8, 20140818 -- Improve dependency management

        * Loosen dependencies in setup.py
        * Fix signature/headers regular expressions
          (Thanks Christopher Grebs)

v0.1.7, 20140802 -- Fix fetching user data

        * Correctly handle a bad API KEY, adapt test code
          (Thanks @benctamas)

v0.1.6, 20140706 -- Slight improvement to API

        * Return API key on successful authentication
          (Thanks Konstantinos Koukopoulos)

v0.1.5, 20140613 -- Document installation issue

        * Document workaround on installation problems

v0.1.4, 20140613 -- Improve installation

        * Make requirements file comply with docs
        * Decide on http_signature commit

v0.1.3, 20140220 -- Upload to PyPI

        * Prepare docs to upload package to PyPI

v0.1.2, 20140219 -- Package data and clean up

        * Updated package classifiers
        * Cleaned up unused code in authentication.py

v0.1.1, 20140217 -- Documentation and clean up

        * The package can be installed
        * Continuous integration via Travis
        * Unit tests for the authentication code
        * General docuementation in the README file

v0.1.0, 20140217 -- Initial release


test_settings.py
DATABASES = {
    'default': {
        'ENGINE': 'django.db.backends.sqlite3',
        'NAME': ':memory:',
    },
}

INSTALLED_APPS = (
    'django.contrib.contenttypes',
    'django.contrib.auth',
    'rest_framework',
    'rest_framework_httpsignature',
)

ROOT_URLCONF = 'rest_framework_httpsignature.tests'

SECRET_KEY = 'MY PRIVATE SECRET'


rest_framework_httpsignature/tests.py
from django.test import SimpleTestCase, TestCase, RequestFactory
from django.contrib.auth import get_user_model
from rest_framework_httpsignature.authentication import SignatureAuthentication
from rest_framework.exceptions import AuthenticationFailed
import re

User = get_user_model()

ENDPOINT = '/api'
METHOD = 'GET'
KEYID = 'some-key'
SECRET = 'my secret string'
SIGNATURE = 'some.signature'


def build_signature(headers, key_id=KEYID, signature=SIGNATURE):
    """Build a signature string."""
    template = ('Signature keyId="%(keyId)s",algorithm="hmac-sha256",'
                'headers="%(headers)s",signature="%(signature)s"')
    return template % {
        'keyId': key_id,
        'signature': signature,
        'headers': ' '.join(headers),
    }


class HeadersUnitTestCase(SimpleTestCase):

    request = RequestFactory()

    def setUp(self):
        self.auth = SignatureAuthentication()

    def test_special_header_names(self):
        for special in ['Content-Type', 'CONTENT-TYPE', 'content-type']:
            canon = self.auth.header_canonical(special)
            self.assertEqual('CONTENT-TYPE', canon)

        for special in ['Content-Length', 'CONTENT-LENGTH', 'content-length']:
            canon = self.auth.header_canonical(special)
            self.assertEqual('CONTENT-LENGTH', canon)

    def test_header_names(self):
        headers = ['X-Api-Key', 'Authentication', 'date', 'X-Something-Else']
        for header in headers:
            canon = self.auth.header_canonical(header)
            expected = 'HTTP_%s' % header.upper().replace('-', '_')
            self.assertEqual(expected, canon)

    def test_build_signature_for_date(self):
        req = self.request.get(ENDPOINT, {}, HTTP_X_DATE="some date")
        dict_to_sign = self.auth.build_dict_to_sign(req, ['date'])
        self.assertTrue('date' in dict_to_sign.keys())

    def test_build_signature_for_date_and_other(self):
        req = self.request.get(ENDPOINT, {}, HTTP_X_ACCEPT="*/*",
                               HTTP_X_DATE="some date")
        dict_to_sign = self.auth.build_dict_to_sign(req, ['accept', 'date'])
        self.assertTrue('date' in dict_to_sign.keys())
        self.assertTrue('accept' in dict_to_sign.keys())

    def test_build_signature_for_request_line(self):
        req = self.request.get(ENDPOINT, {}, HTTP_X_DATE="some date")
        dict_to_sign = self.auth.build_dict_to_sign(
            req,
            ['(request-target)', 'date'])
        self.assertTrue('date' in dict_to_sign.keys())
        self.assertTrue('(request-target)' not in dict_to_sign.keys())


class SignatureTestCase(SimpleTestCase):

    def setUp(self):
        self.auth = SignatureAuthentication()

    def test_no_headers_in_signature_is_date(self):
        signature = build_signature([])
        headers = self.auth.get_headers_from_signature(signature)
        self.assertEqual(1, len(headers))
        self.assertEqual("date", headers[0])

    def test_date_in_signature(self):
        signature = build_signature(['date'])
        headers = self.auth.get_headers_from_signature(signature)
        self.assertTrue('date' in headers)

    def test_many_in_signature(self):
        signature = build_signature(['date', 'accept', '(request-target)'])
        headers = self.auth.get_headers_from_signature(signature)
        self.assertTrue('date' in headers)
        self.assertTrue('accept' in headers)
        self.assertTrue('(request-target)' in headers)

    def test_get_signature(self):
        signature_string = build_signature(['(request-target)', 'date'])
        signature = self.auth.get_signature_from_signature_string(
            signature_string)
        self.assertEqual(SIGNATURE, signature)

    def test_get_signature_without_headers(self):
        signature_string = build_signature([])
        signature = self.auth.get_signature_from_signature_string(
            signature_string)
        self.assertEqual(SIGNATURE, signature)


class BuildSignatureTestCase(SimpleTestCase):

    request = RequestFactory()
    KEYID = 'su-key'

    def setUp(self):
        self.auth = SignatureAuthentication()

    def test_build_signature(self):
        # TO SIGN:
        #
        # GET /packages/measures/ HTTP/1.1
        # host: localhost:8000
        # accept: application/json
        # date: Mon, 17 Feb 2014 06:11:05 GMT

        headers = ['(request-target)', 'host', 'accept', 'date']
        expected_signature = '+dV3yojX7N5I5J+rx0N+7kL5zES2L9Goo4ApJIn33IM='
        expected_signature_string = build_signature(
            headers,
            key_id=self.KEYID,
            signature=expected_signature)

        req = RequestFactory().get(
            '/packages/measures/', {},
            HTTP_HOST='localhost:8000',
            HTTP_ACCEPT='application/json',
            HTTP_DATE='Mon, 17 Feb 2014 06:11:05 GMT',
            HTTP_AUTHORIZATION=expected_signature_string)

        signature_string = self.auth.build_signature(
            self.KEYID, SECRET, req)
        signature = re.match(
            '.*signature="(.+)",?.*', signature_string).group(1)
        self.assertEqual(expected_signature, signature)


class SignatureAuthenticationTestCase(TestCase):

    class APISignatureAuthentication(SignatureAuthentication):
        """Extend the SignatureAuthentication to test it."""

        API_KEY_HEADER = 'X-Api-Key'

        def __init__(self, user):
            self.user = user

        def fetch_user_data(self, api_key):
            if api_key != KEYID:
                return None

            return (self.user, SECRET)

    TEST_USERNAME = 'test-user'
    TEST_PASSWORD = 'test-password'

    def setUp(self):
        self.test_user = User(username=self.TEST_USERNAME)
        self.test_user.set_password(self.TEST_PASSWORD)
        self.auth = self.APISignatureAuthentication(self.test_user)

    def test_no_credentials(self):
        request = RequestFactory().get(ENDPOINT)
        res = self.auth.authenticate(request)
        self.assertIsNone(res)

    def test_only_api_key(self):
        request = RequestFactory().get(
            ENDPOINT, {},
            HTTP_X_API_KEY=KEYID)
        self.assertRaises(AuthenticationFailed,
                          self.auth.authenticate, request)

    def test_bad_signature(self):
        request = RequestFactory().get(
            ENDPOINT, {},
            HTTP_X_API_KEY=KEYID,
            HTTP_AUTHORIZATION='some-wrong-value')
        self.assertRaises(AuthenticationFailed,
                          self.auth.authenticate, request)

    def test_can_authenticate(self):
        headers = ['(request-target)', 'accept', 'date', 'host']
        expected_signature = 'SelruOP39OWoJrSopfYJ99zOLoswmpyGXyDPdebeELc='
        expected_signature_string = build_signature(
            headers,
            key_id=KEYID,
            signature=expected_signature)
        request = RequestFactory().get(
            '/packages/measures/', {},
            HTTP_ACCEPT='application/json',
            HTTP_DATE='Mon, 17 Feb 2014 06:11:05 GMT',
            HTTP_HOST='localhost:8000',
            HTTP_AUTHORIZATION=expected_signature_string,
            HTTP_X_API_KEY=KEYID)

        result = self.auth.authenticate(request)
        self.assertIsNotNone(result)
        self.assertEqual(result[0], self.test_user)
        self.assertEqual(result[1], KEYID)


rest_framework_httpsignature/authentication.py
from rest_framework import authentication
from rest_framework import exceptions
from httpsig import HeaderSigner
import re


class SignatureAuthentication(authentication.BaseAuthentication):

    SIGNATURE_RE = re.compile('signature="(.+?)"')
    SIGNATURE_HEADERS_RE = re.compile('headers="([\(\)\sa-z0-9-]+?)"')

    API_KEY_HEADER = 'X-Api-Key'
    ALGORITHM = 'hmac-sha256'

    def get_signature_from_signature_string(self, signature):
        """Return the signature from the signature header or None."""
        match = self.SIGNATURE_RE.search(signature)
        if not match:
            return None
        return match.group(1)

    def get_headers_from_signature(self, signature):
        """Returns a list of headers fields to sign.

        According to http://tools.ietf.org/html/draft-cavage-http-signatures-03
        section 2.1.3, the headers are optional. If not specified, the single
        value of "Date" must be used.
        """
        match = self.SIGNATURE_HEADERS_RE.search(signature)
        if not match:
            return ['date']
        headers_string = match.group(1)
        return headers_string.split()

    def header_canonical(self, header_name):
        """Translate HTTP headers to Django header names."""
        # Translate as stated in the docs:
        # https://docs.djangoproject.com/en/1.6/ref/request-response/#django.http.HttpRequest.META
        header_name = header_name.lower()
        if header_name == 'content-type':
            return 'CONTENT-TYPE'
        elif header_name == 'content-length':
            return 'CONTENT-LENGTH'
        return 'HTTP_%s' % header_name.replace('-', '_').upper()

    def build_dict_to_sign(self, request, signature_headers):
        """Build a dict with headers and values used in the signature.

        "signature_headers" is a list of lowercase header names.
        """
        d = {}
        for header in signature_headers:
            if header == '(request-target)':
                continue
            d[header] = request.META.get(self.header_canonical(header))
        return d

    def build_signature(self, user_api_key, user_secret, request):
        """Return the signature for the request."""
        path = request.get_full_path()
        sent_signature = request.META.get(
            self.header_canonical('Authorization'))
        signature_headers = self.get_headers_from_signature(sent_signature)
        unsigned = self.build_dict_to_sign(request, signature_headers)

        # Sign string and compare.
        signer = HeaderSigner(
            key_id=user_api_key, secret=user_secret,
            headers=signature_headers, algorithm=self.ALGORITHM)
        signed = signer.sign(unsigned, method=request.method, path=path)
        return signed['authorization']

    def fetch_user_data(self, api_key):
        """Retuns (User instance, API Secret) or None if api_key is bad."""
        return None

    def authenticate(self, request):
        # Check for API key header.
        api_key_header = self.header_canonical(self.API_KEY_HEADER)
        api_key = request.META.get(api_key_header)
        if not api_key:
            return None

        # Check if request has a "Signature" request header.
        authorization_header = self.header_canonical('Authorization')
        sent_string = request.META.get(authorization_header)
        if not sent_string:
            raise exceptions.AuthenticationFailed('No signature provided')
        sent_signature = self.get_signature_from_signature_string(sent_string)

        # Fetch credentials for API key from the data store.
        try:
            user, secret = self.fetch_user_data(api_key)
        except TypeError:
            raise exceptions.AuthenticationFailed('Bad API key')

        # Build string to sign from "headers" part of Signature value.
        computed_string = self.build_signature(api_key, secret, request)
        computed_signature = self.get_signature_from_signature_string(
            computed_string)

        if computed_signature != sent_signature:
            raise exceptions.AuthenticationFailed('Bad signature')

        return (user, api_key)


utils/sign3.py
import argparse
import hashlib
import base64
import hmac


def parse_args():
    """Parse program arguments into an object."""
    parser = argparse.ArgumentParser()
    parser.add_argument('key', help='Key to sign with')
    parser.add_argument('key_id', help='an ID of the key in use')
    return parser.parse_args()


def string_to_sign():
    return '\n'.join([
        '(request-target): get /packages/measures/',
        'accept: application/json',
        'date: Mon, 17 Feb 2014 06:11:05 GMT',
        'host: localhost:8000'])


def raw_sign(message, secret):
    """Sign a message."""
    digest = hmac.new(secret, message, hashlib.sha256).digest()
    return base64.b64encode(digest)


def http_signature(message, key_id, signature):
    """Return a tuple (message signature, HTTP header message signature)."""
    template = ('Signature keyId="%(keyId)s",algorithm="hmac-sha256",'
                'headers="%(headers)s",signature="%(signature)s"')
    headers = ['(request-target)', 'host', 'accept', 'date']
    return template % {
        'keyId': key_id,
        'signature': signature,
        'headers': ' '.join(headers),
    }


if __name__ == '__main__':
    args = parse_args()
    message = string_to_sign()
    signature = raw_sign(message, args.key)
    header = http_signature(message, args.key_id, signature)
    print "=== SIGNATURE ==="
    print signature
    print "=== HTTP HEADER SIGNATURE ==="
    print header
